pkg/operator/ceph/cluster/key_rotation.go

```diff
@@ -0,0 +1,102 @@
+limitations under the License.
+*/
+
+// Package cluster to manage a Ceph cluster.
+package cluster
+
+import (
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/ceph/config"
+	"github.com/rook/rook/pkg/operator/ceph/config/keyring"
+	"github.com/rook/rook/pkg/operator/ceph/reporting"
+)
+
+// ClusterKeysToRotate defines the type of ceph daemon keys that need rotation
+type ClusterKeysToRotate struct {
+	RBDMirrorPeer bool
+}
+
+// GetKeysToRotate returns the KeysToRotate struct that defines the type of keys that need rotation.
+func (c *cluster) GetKeysToRotate() (ClusterKeysToRotate, error) {
+	keysToRotate := ClusterKeysToRotate{}
+	cephObj := &cephv1.CephCluster{}
+	if err := c.context.Client.Get(c.ClusterInfo.Context, c.namespacedName, cephObj); err != nil {
+		return keysToRotate, errors.Wrapf(err, "failed to get cluster %v to get the cephx keys to rotate.", c.namespacedName)
+	}
+
+	// cephX status won't be initialized for a new cluster.
+	if cephObj.Status.CephX == nil {
+		return keysToRotate, nil
+	}
+
+	desiredCephVersion := c.ClusterInfo.CephVersion
+	// check the least version for OSD since mons might have been upgraded by now.
+	runningCephVersion, err := client.LeastUptodateDaemonVersion(c.context, c.ClusterInfo, config.MonType)
+	if err != nil {
+		return keysToRotate, errors.Wrapf(err, "failed to get the lowest running ceph version in the cluster %v.", c.namespacedName)
+	}
+
+	allCephDaemonTypes := []string{client.RBDMirrorPeerKeyringID}
+	for _, daemon := range allCephDaemonTypes {
+		switch daemon {
+		case client.RBDMirrorPeerKeyringID:
+			shouldRotateKeys, err := keyring.ShouldRotateCephxKeys(cephObj.Spec.Security.CephX.RBDMirrorPeer, runningCephVersion, desiredCephVersion, *cephObj.Status.CephX.RBDMirrorPeer)
+			if err != nil {
+				return keysToRotate, errors.Wrapf(err, "failed to check if %q daemon keys should be rotated or not", daemon)
+			}
+			keysToRotate.RBDMirrorPeer = shouldRotateKeys
+		}
+	}
+
+	logger.Debugf("ceph daemon keys to rotate in cluster %v are: %v", c.namespacedName, keysToRotate)
+	return keysToRotate, nil
+}
+
+// UpdateClusterCephxStatus fetches the latest cephCluster instance and updates cephxStatus for a particular ceph daemon
+func (c *cluster) UpdateClusterCephxStatus(cephxDaemonType string, didRotate bool) error {
+	cluster := &cephv1.CephCluster{}
+	if err := c.context.Client.Get(c.ClusterInfo.Context, c.namespacedName, cluster); err != nil {
+		return errors.Wrapf(err, "failed to get cluster %v to update the conditions.", c.namespacedName)
+	}
+
+	if cluster.Status.CephX == nil {
+		cluster.Status.CephX = &cephv1.ClusterCephxStatus{}
+	}
+
+	switch cephxDaemonType {
+	case client.RBDMirrorPeerKeyringID:
+		if cluster.Status.CephX.RBDMirrorPeer == nil {
+			uninitializedStatus := keyring.UninitializedCephxStatus()
```

Comment:
As discussed in huddle, let's set `Uninitialized` for all `cluster.status.cephx` subsections at the same time early in the CephCluster reconcile.